en/docs/parallelism/03_consistent_tensor.md

```diff
@@ -0,0 +1,244 @@
+    flow.Size([4, 8])
+    ```
+
+### Model-Parallelism
+
+The following code is an example of [model-parallelism of common distributed parallelism](./01_introduction.md#_5).
+
+![data parallelism](./imgs/matmul_model_paralelism.png)
+
+=== "Terminal 0"
+    ```python
+    import oneflow as flow
+
+    placement = flow.placement("cuda",{0:[0,1]})
+    x = flow.randn(4,5,placement=placement, sbp=flow.sbp.broadcast)
+    w = flow.randn(5,8,placement=placement, sbp=flow.sbp.split(1))
+    y = flow.matmul(x,w)
+    y.sbp
+    y.shape
+    ```
+
+=== "Terminal 1"
+    ```python
+    import oneflow as flow
+
+    placement = flow.placement("cuda",{0:[0,1]})
+    x = flow.randn(4,5,placement=placement, sbp=flow.sbp.broadcast)
+    w = flow.randn(5,8,placement=placement, sbp=flow.sbp.split(1))
+    y = flow.matmul(x,w)
+    y.sbp
+    y.shape
+    ```
+
+Observe that `flow.matmul` checks its input `x` and `w` whose SBP are `broadcast` and `split(0)`, respectively. Oneflow then derives and outputs the SBP of `y` which is `split(1)`. In the end, computation is done with a matrix of `shape=(4,8)`. Output:
+
+=== "Terminal 0"
+    ```text
+    (oneflow.sbp.split(axis=1),)
+    flow.Size([4, 8])
+    ```
+
+=== "Terminal 1"
+    ```text
+    (oneflow.sbp.split(axis=1),)
+    flow.Size([4, 8])
+    ```
+
+## Extended Reading
+
+### Environment Variables in Multi-Machine Training
+
+The example in this article sets environment variables to configure distributed training. Doing so allows programmers to see the effects and output in an interactive python environment. If the training is needed in production instead of in learning or experiments, one may launch distributed training using [oneflow.distributed.launch](./04_launch.md). This module automatically sets necessary environment variables based on command-line arguments.
+
+- `MASTER_ADDR`：The IP of the 0th machine in a multi-machine case
+- `MASTER_PORT`：The listening port of the 0th machine in a multi-machine case. Note that this port should not be in use
+- `WORLD_SIZE`：The number of computing devices in the whole cluster. Because currently oneflow only supports having the  same number of GPUs on each machines, `WORLD_SIZE` is actually $number\:of\:machines \times number\:of\:GPUs\:on\:one\:machine$. In our example, we have one machine and two GPUs on it, so `WORLD_SIZE=2`
+
+`RANK` and `LOCAL_RANK` are indexes for machines. The difference is that `RANK` is a "global perspective" index, while `LOCAL_RANK` is a "local perspective" index. In the case that only one machine is involved, the `RANK` and `LOCAL_RANK` are the same. In our example, there are two GPUs, indexed 0 and 1.
+
+When there are multiple machines, the upper bound of `LOCAL_RANK` on a machine is the number of computing devices on the machine. The upper bound of `RANK` is the sum of all computing devices on all machines. The indexing of these computing devices starts from 0. (Both upper bounds are non-inclusive since indexing starts from 0)
```

Comment:
When launching the distributed training on multiple machines, the upper bound of `LOCAL_RANK` keeps the same with the number of computing devices on single machine. The upper bound of `RANK` keeps the same with the total number of computing devices in the cluster. The indexing of processes starts from 0. (Both upper bounds are non-inclusive since indexing starts from 0)

Reply:
已修改